Cleanup TransportReplicationAction

martijnvg opened the pull request:
* Split the actual primary operation from the primary phase into a dedicated AsyncPrimaryAction class (similar to AsyncReplicaAction)
* Removed threaded_operation option from replication based transport actions.
* Let the threading be handled by by the transport service and drop forking from new threads from the transport replication action.

brwe commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -881,68 +922,25 @@ void performOnReplica(final ShardRouting shard, final String nodeId) {
-
-                        });
-            } else {
-                if (replicaRequest.operationThreaded()) {
```
so happy that you remove this part!

brwe commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -309,8 +543,8 @@ protected void responseWithFailure(Throwable t) {
         protected void doRun() throws Exception {
-            try (Releasable shardReference = getIndexShardOperationsCounter(request.internalShardId)) {
-                shardOperationOnReplica(request.internalShardId, request);
+            try (Releasable shardReference = getIndexShardOperationsCounter(request.internalShardRouting.shardId())) {
```
This breaks the `ShardReplicationTests.testReplicasCounter()` at least it hangs now. This is not a problem with this pr but more with the test - it needs the internalShardRouting now to be initialized, else we get an NPE in the test.

martijnvg replied:
I noticed that too but eventually it passes, but I somehow thought this was also the case without this change.

martijnvg replied:
@brwe so I'll change the test to do this.

brwe replied:
:+1:

martijnvg commented:
@brwe I've fixed the test and made sure that the newly added `ReplicationRequest#internalShardRouting` is always used in the test.

martijnvg commented:
I rebased this PR and moved the retry primary handling entirely to the coordinating node. 

Also I included @brwe test from #12574 which passes with this PR. This test simulates a situation where 2 nodes endlessly redirect same index request to each other caused by a number of events described in #12573.

The reason this test is included is because the way primary write requests are redirected has changed in the PR. The coordinating node remains in charge of the entire write operation, even if the primary shard is on a different node than the coordinating node. While before the node that holds the primary shard is always in charge of a write request. In a situation that is described in #12573 this can lead to nodes endlessly redirecting the same write request to each other until the cluster state caught up or something bad happens. With this PR with the #12573 situation, the write request will be retried when a new cluster state comes or the write timeout has been met.

brwe commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -242,10 +241,177 @@ public void messageReceived(final ReplicaRequest request, final TransportChannel
+                    try {
+                        channel.sendResponse(e);
+                    } catch (Throwable t) {
+                        logger.warn("failed to send response for get", t);
```
for write request?

martijnvg replied:
yes :)

brwe commented on core/src/test/java/org/elasticsearch/discovery/EndlessIndexingLoopIT.java:
```diff
@@ -0,0 +1,181 @@
+                .setWaitForRelocatingShards(0)
+                .get();
+        assertThat(clusterHealth.isTimedOut(), equalTo(false));
+        indexResponseFuture.get();
```
hm. so this hangs now every now and then for a minute. I think it is when the coordinating node is node_1. Then the cluster state observer waits for the next cluster state which does not come and the index request is only executed when the observer times out.
We can send the request via node_2 but I think we actually need a way to handle this better.

martijnvg replied:
I'll test this out. I've seen this hanging too.

martijnvg replied:
I've changed it to use the client from node `node_2` we can also reduce the timeout to `10s` or something similar?

brwe commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -242,10 +241,177 @@ public void messageReceived(final ReplicaRequest request, final TransportChannel
+
+        @Override
+        public void onFailure(Throwable t) {
+            finishAsFailed(t);
```
I am not sure anymore if we can completely remove the retry mechanism from the PrimaryAction. Here is my reasoning, let me know if it makes sense: If we only retry on the coordinating node than we assume that the coordinating node is not up-to-date and might receive a newer cluster state after the failure. However, it might also be for a very short time that the receiving node that has the primary is not up-to-date. This is the case in the test for the endless request loop. In this case the coordinating node might not receive a new cluster state at all and the indexing request would then only be executed when the observer times out. The coordinating node has currently no way of knowing if the receiving node got a new cluster state or not. If we had a retry in the AsyncPrimaryAction too we could catch that maybe?

martijnvg replied:
Right, I see what you mean. This is a reason to put back the retry mechanism in `AsyncPrimaryAction`. Otherwise we would delay the write request for no reason, which isn't what we want.

But on the other hand then we have 2 retry mechanisms and we then also need a way of preventing that we retry twice. But some errors we really want only to retry on the coordination node, for example, if a the primary shard has just been removed on the node that did hold the primary and the primary shard is now somewhere else. In that case it is best to retry from the coordination node. So we need to properly deal with these kinds of situations.

martijnvg replied:
Just chatted with @brwe and delaying the write request upto the timeout isn't the end of the world in this situation and this situation is rare too. It is much better then the current behaviour. (redirecting the write request endlessly between nodes until the timeout expired)

brwe commented on core/src/test/java/org/elasticsearch/action/support/replication/ShardReplicationTests.java:
```diff
@@ -314,11 +352,10 @@ public void testRoutingToPrimary() {
-            assertThat(capturedRequests.size(), equalTo(1));
-            assertThat(capturedRequests.get(0).action, equalTo("testAction"));
+            logger.info("--> primary is assigned to [{}], checking request is going to be retried at some point", primaryNodeId);
+            assertThat(clusterService.getListeners().size(), equalTo(0));
```
I don't understand why we would expect a retry here. Cluster state is fine, this should just send the request.  Also, why remove the check for captured requests?

martijnvg replied:
hmm... this looks like it needs to be reverted.

martijnvg replied:
now I remember why... I added a comment

brwe commented on core/src/test/java/org/elasticsearch/action/support/replication/ShardReplicationTests.java:
```diff
@@ -304,9 +342,9 @@ public void testRoutingToPrimary() {
-        primaryPhase.routeRequestOrPerformLocally(shardRoutingTable.primaryShard(), shardRoutingTable.shardsIt());
         if (primaryNodeId.equals(clusterService.localNode().id())) {
             logger.info("--> primary is assigned locally, testing for execution");
+            primaryPhase.moveToPrimaryAction(shardRoutingTable.primaryShard());
```
Can we somehow split testing of the three phases? For me the transport handler that executes requests and also the AsyncPrimaryAction makes it more difficult to understand which part of TransportReplicationAction is tested.  Or is it very hard to split that?

martijnvg replied:
not sure if it is too hard, I'm willing to try it out.

brwe commented:
Overall I really like this split of primary operation and primary phase. I think this might make it much easier to understand what is happening. I left some comments and questions around testing and where we should retry though. Hope they are not too confusing...

martijnvg commented:
@bleskes I brought back the chasing of the primary shard to `PrimaryPhase` inner class and delegate to the primary action if the primary shard is local.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -242,10 +241,179 @@ public void messageReceived(final ReplicaRequest request, final TransportChannel
+            final ShardRouting primary = request.internalShardRouting;
+            // Although this gets executed locally, this more of an assertion, but if change the primary action
+            // to be performed remotely this check is important to check before performing the action:
+            if (clusterService.localNode().id().equals(primary.currentNodeId()) == false) {
```
do we need to do this? if we fail to find the shard in the indicesService, we'll throw an exception anyhow?

martijnvg replied:
true. it kind of made sense how the pr did work, i left it because imo it would be a nice check. but lets just remove it.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -41,12 +41,11 @@
     public static final TimeValue DEFAULT_TIMEOUT = new TimeValue(1, TimeUnit.MINUTES);
 
-    ShardId internalShardId;
+    ShardRouting internalShardRouting;
```
do we still need this change? see other comment on transport replication action

martijnvg replied:
the `AsyncPrimaryAction` is using this now.  We can remove this but then we need to resolve the primary twice. I don't have a real preference, since it is all local now.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -41,12 +41,11 @@
     protected TimeValue timeout = DEFAULT_TIMEOUT;
     protected String index;
 
-    private boolean threadedOperation = true;
```
w00t :)

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -102,11 +101,13 @@ protected TransportReplicationAction(Settings settings, String actionName, Trans
         this.checkWriteConsistency = checkWriteConsistency();
 
         transportService.registerRequestHandler(actionName, request, ThreadPool.Names.SAME, new OperationTransportHandler());
+        transportService.registerRequestHandler(transportPrimaryAction, request, executor, new PrimaryOperationTransportHandler());
```
can we add java docs somewhere (on the handlers, or on the top of  this class explaining) what the responsibility of each handler is?

martijnvg replied:
+1

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -348,8 +533,6 @@ public RetryOnPrimaryException(StreamInput in) throws IOException{
         private final InternalRequest internalRequest;
```
This the first line I can comment on for this class- can we update the java doc (it doesn't do everything up to sending requests to replicas any more). 

Also in terms of naming: we used to have PrimaryPhase and ReplicaPhase  , now some of the logic is split between PrimaryPhase , AsyncPrimaryAction and ReplicaPhase.   I think it would be clearer if we have ResolvePrimaryPhase, PrimaryPhase (which is the logic in AsyncPrimaryAction) and ReplicaPhase.

martijnvg commented:
The tricky part about the PR is that the cluster state is observed twice, one time in the primary phase and one time in the async primary action. In cases we retry, we might miss the update the to the cluster state. This would only be likely to occur more if the primary action was executed remotely which, is what the plan was after as follow up issue. But this can be dangerous. It is fixable if we remember on what version we decided to execute the primary operation, but that would make this change bigger than the plan is and we should maybe to a break from this change and reconsider it post 2.0
